x/denommetadata/ibc_middleware.go

```diff
@@ -21,89 +20,77 @@ var _ porttypes.IBCModule = &IBCMiddleware{}
 // IBCMiddleware implements the ICS26 callbacks for the transfer middleware
 type IBCMiddleware struct {
 	porttypes.IBCModule
-	keeper DenomMetadataKeeper
+	keeper Keeper
 }
 
-type DenomMetadataKeeper interface {
+type Keeper interface {
 	HasDenomMetaData(ctx sdk.Context, denom string) bool
 	CreateDenomMetadata(ctx sdk.Context, metadatas ...types.DenomMetadata) error
 }
 
 // NewIBCMiddleware creates a new IBCMiddleware given the keeper and underlying application
-func NewIBCMiddleware(keeper DenomMetadataKeeper, app porttypes.IBCModule) IBCMiddleware {
+func NewIBCMiddleware(keeper Keeper, app porttypes.IBCModule) IBCMiddleware {
 	return IBCMiddleware{
 		IBCModule: app,
 		keeper:    keeper,
 	}
 }
 
-// OnRecvPacket registers the denom metadata if it does not exist
+// OnRecvPacket registers the denom metadata if it does not exist.
+// It will intercept an incoming packet and check if the denom metadata exists.
+// If it does not, it will register the denom metadata.
+// The handler will expect a 'transferinject' object in the memo field of the packet.
+// If the memo is not an object, or does not contain the metadata, it moves on to the next handler.
 func (im IBCMiddleware) OnRecvPacket(
 	ctx sdk.Context,
 	packet channeltypes.Packet,
 	relayer sdk.AccAddress,
 ) exported.Acknowledgement {
 	packetData := new(transfertypes.FungibleTokenPacketData)
 	if err := types.ModuleCdc.UnmarshalJSON(packet.GetData(), packetData); err != nil {
-		err = errorsmod.Wrapf(errortypes.ErrInvalidType, "cannot unmarshal ICS-20 transfer packet data")
+		err = errorsmod.Wrapf(errortypes.ErrJSONUnmarshal, "unmarshal ICS-20 transfer packet data")
 		return channeltypes.NewErrorAcknowledgement(err)
 	}
 
-	if len(packetData.Memo) == 0 {
+	transferInject, err := types.ParsePacketMetadata(packetData.Memo)
+	if err != nil {
 		return im.IBCModule.OnRecvPacket(ctx, packet, relayer)
 	}
 
-	memoData, err := types.ParseMemoData(packetData.Memo)
-	// if the memo is not an object, or does not contain the metadata, we can skip
-	if errors.Is(err, types.ErrMemoUnmarshal) || errors.Is(err, types.ErrMemoDenomMetadataEmpty) {
-		return im.IBCModule.OnRecvPacket(ctx, packet, relayer)
-	}
+	packetData.Memo = types.PurgePacketMetadata(packetData.Memo)
+	// re-marshal the packet data
+	packet.Data, err = types.ModuleCdc.MarshalJSON(packetData)
 	if err != nil {
 		return channeltypes.NewErrorAcknowledgement(err)
 	}
```

Comment:
why purge?
Isn't that going to break things?
Pretty sure this will break acks
Did you test the whole thing locally with running chain and RA?

Reply:
it just removes the `transferinject` stuff from the memo